### DISCUSSION TITLE
Caching

### DISCUSSION THREAD

**Discussion by BMarques** [Unidata/tds#549] [General]
Hello,
I'm trying to understand how cache works in Thredds 5.5, and failing to do so, so far.

We have a thredds server online (running in ubuntu 22.04) and working for about 6 months without issues, but looking at the /thredds/cache folder, there's no single file there, only the predefined cache folders, and our custom configured folder is also empty.

We mainly use Thredds to serve files with WMS protocol to our web platform, but we also make sure of the WCS protocol for water streamlines. 

Since most of our files are not in the thredds server, but rather in our NAS and other servers (through CIFS and NFS), I was hoping to be able to cache some of those files that are more regularly accessed, so I set up the following in threddsConfig.xml

```
<DiskCache>
    <alwaysUse>true</alwaysUse>
    <dir>/threddsCache</dir>
    <scour>1 hour</scour>
    <maxSize>10 Gb</maxSize>
  </DiskCache> 
```

```
  <WCS>
    <allow>true</allow>
    <allowRemote>false</allowRemote>
    <scour>15 min</scour>
    <maxAge>30 min</maxAge>
    <dir>/threddsCache/</dir>
  </WCS>
```

Nevertheless, no files end up in that /threddsCache folder and the cache.logs don't say anything helpful
   
```
[2024-11-28T11:55:39.850+0000] Started a DiskCache2 scour task on /data/content/thredds/cache/agg every 1440 minutes for files older than 129600 minutes
[2024-11-28T11:55:39.850+0000] Started a DiskCache2 scour task on /data/content/thredds/cache/aggNew/ every 1440 minutes for files older than 129600 minutes
[2024-11-28T11:55:39.856+0000] Started a DiskCache2 scour task on /data/content/thredds/cache/ncss every 15 minutes for files older than 30 minutes
[2024-11-28T11:55:41.543+0000] Started a DiskCache2 scour task on /threddsCache/ every 15 minutes for files older than 30 minutes
```

Regarding WMS caching, I don't know if there's any configuration to be done in the WMS cache, but I'm also failing to understand its behaviour, it definitely feels like it's working for some time, maybe it's saved in memory?

So yeah, I don't know if it's working, or where to look at because anything i'm doing, seems to simply do nothing, hence I started this discussion.

**Comment by lesserwhirls** [comment]
We have some documentation about the various caches used by the TDS in our documentation regarding the `threddsConfig.xml` file: https://docs.unidata.ucar.edu/tds/current/userguide/tds_config_ref.html#disk-caching-and-temporary-files

Caching in the WMS service heavily relies on a mixture of existing in-memory object caches (file handles, `NetcdfDataset` objects, NcML aggregations, etc.) and are not necessarily persisted between service restarts (it depends on the cache). The WMS specific cache is in-memory (maximum size set to 100 entries (roughly speaking, datasets)), isn't persisted, and does not have any configuration options exposed at this point (although that would useful enhancement).

**Reply by BMarques** [reply]
Hello,
I have read that documentation previously, but still it feels like no cache is being done.
Thanks for the details on the WMS cache.
